core/src/scene/drawRule.cpp

```diff
@@ -1,120 +1,152 @@
+                    StyleContext& _ctx, Tile& _tile) {
 
-bool DrawRule::operator<(const DrawRule& _rhs) const {
-    return name < _rhs.name;
-}
+    if (!match(_feature, _layer, _ctx)) { return; }
 
-bool DrawRule::eval(const StyleContext& _ctx) {
+    // Apply styles
+    for (auto& rule : styles) {
+
+        auto& styleName = rule.getStyleName();
+        int styleId = styleName.empty()
+            ? rule.styleId
+            : _scene.getStyleId(styleName);
+
+        if (styleId < 0){
+            LOGE("Invalid style %s", rule.getStyleName().c_str());
+            continue;
+        }
 
-    for (auto& param : parameters) {
-        if (param.function >= 0) {
-            if (!_ctx.evalStyle(param.function, param.key, param.value)) {
-                if (StyleParam::isRequired(param.key)){
-                    return false;
+        // Evaluate JS functions and Stops
+        bool valid = true;
+        for (size_t i = 0; i < StyleParamKeySize; ++i) {
+            auto* param = rule.params[i];
+            if (!param) { continue; }
+
+            if (param->function >= 0) {
+                if (!_ctx.evalStyle(param->function, param->key, rule.evaluated[i].value) &&
+                    StyleParam::isRequired(param->key)) {
+                    valid = false;
+                    break;
                 }
             }
-        }
-        if (param.stops) {
-            if (StyleParam::isColor(param.key)) {
-                param.value = param.stops->evalColor(_ctx.getGlobalZoom());
-            } else if (StyleParam::isWidth(param.key)) {
-                param.value = param.stops->evalWidth(_ctx.getGlobalZoom());
-            } else {
-                param.value = param.stops->evalFloat(_ctx.getGlobalZoom());
+            if (param->stops) {
+                rule.evaluated[i].stops = param->stops;
+
+                if (StyleParam::isColor(param->key)) {
+                    rule.evaluated[i].value = param->stops->evalColor(_ctx.getGlobalZoom());
+                } else if (StyleParam::isWidth(param->key)) {
+                    // FIXME widht result is isgnored from here..
+                    rule.evaluated[i].value = param->stops->evalWidth(_ctx.getGlobalZoom());
+                } else {
+                    rule.evaluated[i].value = param->stops->evalFloat(_ctx.getGlobalZoom());
+                }
             }
+            rule.evaluated[i].key = param->key;
+        }
+
+        if (valid) {
+            auto& style = _scene.styles()[styleId];
+            style->buildFeature(_tile, _feature, rule);
         }
     }
-    return true;
 }
 
-const std::string& DrawRule::getStyleName() const {
+void Styling::mergeRules(const std::vector<StaticDrawRule>& rules) {
+    for (auto& rule : rules) {
 
-    const auto& style = findParameter(StyleParamKey::style);
+        auto it = std::find_if(styles.begin(), styles.end(), [&](auto& m) {
+                return rule.styleId == m.styleId; });
```

Comment:
This confuses me. It looks from the code as if rules are being merged according to their `style`, but they *should* be merged by the names of the rules (which can be the same as their `style`, but in general are not). 

The complete rule application process seems to produce the correct behavior somehow (at least in some brief tests), but at a minimum I'd like to refactor this function to make the intent more clear.